### ISSUE TITLE
Wiki: find and deprecate all unbalanced ability macros

### ISSUE THREAD

**Issue by stevecotton** [wesnoth/wesnoth#7376]
### Describe the desired feature

We still have some unbalanced ability macros on the wiki that are documented as examples, rather than marked as workarounds that were written before better methods were available (or were written after, because the examples weren't updated).

"Unbalanced" might itself be the wrong name for these - they're ones that know they're already in an `[attack][specials]` block, and so can (and do) use `[/attack][/specials]..[+attack][+specials]`.

Checklist for clearing them up:

https://wiki.wesnoth.org/WML_Abilities
- [ ] Pickpocket
- [ ] Soultaker
- [ ] Soultaker (weapon special)
- [ ] Charm (Type 2)

**Comment by ProditorMagnus** [comment]
I dont think there is better way. As far as I know [specials] doesnt support [event] still.

**Comment by Hejnewar** [comment]
From my perspective it would be great if both [specials] and [effect] supported [event] tag. 

Honestly I kinda feel like the only correct way of writing abilities or specials existing thru lua is not very beginner friendly.

**Comment by MrTitainin** [comment]
More than [effect] tag itself (which could be a fine addition as well), it would be better if [abilities] and [set_specials] child tags of [effect] would support it (in addition to normal [specials] and [abilities] in unit of course). This would greatly simplify many UMC abilities and specials and allow much easier portability and reuse.

**Comment by stevecotton** [comment]
> Honestly I kinda feel like the only correct way of writing abilities or specials existing thru lua is not very beginner friendly.

It doesn't (AFAIK, haven't coded one to check) have to be in Lua, it's just that @CelticMinstrel used that when porting the Merman Brawler.

**Comment by stevecotton** [comment]
@CelticMinstrel clarified in Discord that the unbalanced macros aren't actually deprecated.

**Comment by CelticMinstrel** [comment]
I do agree though that supporting `[event]` in specials and abilities would be very convenient.

**Comment by gfgtdf** [comment]
When implementing this, we need to be careful, not to mess up the order in which the events are added especially on save/load. Since the outcome of wml events might depend on the order they are executed in.

**Comment by CelticMinstrel** [comment]
I'd say the second option.

**Comment by gfgtdf** [comment]
Oh and another possible problem might be undoing: When a unit with with a custom  ability is recruited and the recruitment is undone the event might remain in the gamestate possibly leading to oos (probably not if the event filters always on the unit ability though).

**Comment by ProditorMagnus** [comment]
That is no different from [event] in [unit_type].

**Comment by CelticMinstrel** [comment]
It's equivalent to [event] in [unit_type], yes. The event handlers would be registered at the same time as events in the [unit_type]. However, it has the advantage that it can be used without requiring unbalanced tags, which also means it can be used in an [object]. (We'd have to make sure applying the object correctly registers the event handlers, of course.)

**Comment by CelticMinstrel** [comment]
> When a unit with with a custom ability is recruited and the recruitment is undone the event might remain in the gamestate possibly leading to oos (probably not if the event filters always on the unit ability though).

One possible solution to that might be to inject filters into the events. Basically, if there's no `[filter]` tag in the event, add one that contains `type=this type name`, and if there is one, add an `[and]` subtag with the same content. You'd also have to modify the ID so that two different unit types can use the same ability.

That said, I don't think all that trouble is essential. Someone writing events for abilities should already be filtering using `ability_type_id` or `ability_type` in my opinion; if they don't, then they've messed something up and the out-of-sync errors after undoing a recruit are one undesirable result of that mistake (but surely not the only one).

**Comment by gfgtdf** [comment]
> Honestly I kinda feel like the only correct way of writing abilities or specials existing thru lua is not very beginner friendly.

This is not really true, the lua events also need to be registered somewhere, usually in [campaign], [era] or [resource] and of course you can put [event]s there aswell to enable your custom abilities. So here there is no real difference between wml and lua powered abilities right now.

**Comment by CelticMinstrel** [comment]
Even if you're registering the events in Lua, that code needs to go _somewhere_, so basically you just have [lua] instead of [event] in your add-on tag.

**Comment by gfgtdf** [comment]
That's what I just said? Not sure whether misunderstood me or whether you are trying to correct possible spelling mistakes.

**Comment by stevecotton** [comment]
> One possible solution to that might be to inject filters into the events. Basically, if there's no `[filter]` tag in the event, add one that contains `type=this type name`, and if there is one, add an `[and]` subtag with the same content. You'd also have to modify the ID so that two different unit types can use the same ability.

Events might need to trigger on other unit types though. An example would be `moveto` events that put any unit next to the unit with the ability.

**Comment by ProditorMagnus** [comment]
Undo and OOS concerns are not relevant for this discussion. If you recruit unit with [event] in [unit_type] and try to undo it, either something breaks and it is existing bug, or nothing breaks and nothing would break with [event] in [unit_type][abilities][dummy] either.

**Comment by CelticMinstrel** [comment]
> Events might need to trigger on other unit types though. An example would be `moveto` events that put any unit next to the unit with the ability.

Good point. Probably better just to rely on the add-on developer to properly limit it so that the event never triggers if no unit with the ability exists anymore.
